### PR TITLE
Fix an issue ALPN handshake test does not respect "HTTPS_PROXY"

### LINKED ISSUE #27397: HTTPS_PROXY not being correctly used for connections in v13
Expected behavior: If `HTTPS_PROXY` is set in Teleport's environment, all Teleport connections should be HTTP `CONNECT` requests sent via the proxy.

Current behavior: `HTTPS_PROXY` is not used when checking whether an ALPN connection upgrade is required (and potentially for the actual connections themselves)

```
Jun 05 19:05:37 server1 teleport[45820]: 2023-06-05T19:05:37+02:00 DEBU [PROC:1]    Retrying connection to auth server after waiting 3.885913434s. pid:45820.1 service/connect.go:135
Jun 05 19:05:33 server1 teleport[45820]: 2023-06-05T19:05:33+02:00 ERRO [PROC:1]    Instance failed to establish connection to cluster: Failed to connect to Proxy Server through tunnel: connection error: desc = "transport: Error wh>
Jun 05 19:05:33 server1 teleport[45820]: 2023-06-05T19:05:33+02:00 DEBU [HTTP:PROX] Found proxy "http://squid.proxy.local:8080/" in environment, returning proxy dialer. proxy/proxy.go:201
Jun 05 19:05:33 server1 teleport[45820]: 2023-06-05T19:05:33+02:00 INFO             ALPN connection upgrade test failed for "teleport.cluster.com:443": dial tcp 1.1.1.1:443: i/o timeout. client/alpn_conn_upgrade.go:73
Jun 05 19:05:33 server1 teleport[45820]: 2023-06-05T19:05:33+02:00 ERRO [PROC:1]    Node failed to establish connection to cluster: Failed to connect to Proxy Server through tunnel: connection error: desc = "transport: Error while >
Jun 05 19:05:33 server1 teleport[45820]: 2023-06-05T19:05:33+02:00 DEBU [HTTP:PROX] Found proxy "http://squid.proxy.local:8080/" in environment, returning proxy dialer. proxy/proxy.go:201
Jun 05 19:05:33 server1 teleport[45820]: 2023-06-05T19:05:33+02:00 INFO             ALPN connection upgrade test failed for "teleport.cluster.com:443": dial tcp 1.1.1.1:443: i/o timeout. client/alpn_conn_upgrade.go:73
Jun 05 19:05:03 server1 teleport[45820]: 2023-06-05T19:05:03+02:00 DEBU             Attempting GET teleport.cluster.com:443/webapi/find webclient/webclient.go:128
```

Bug details:
- Teleport version: 13.0.3

### PR DESCRIPTION
**Pull request by greedy52**
Fixes #27397 

Related issue

- #27397
- #27461
- #21870

Note 1: [httpproxy.FromEnvironment](https://pkg.go.dev/golang.org/x/net/http/httpproxy#FromEnvironment) skips `localhost` and loopback IPs. Thus decided to overwrite it with a `DialOption` in UT.  
Note 2: I added a utility program but not sure what's the best place to put it. I see we have another utility in `lib/auth/webauthn` though.

Manual testing:
- machine 1: teleport cluster `teleport.dev.aws.stevexin.me`, private network
- machine 2: runs `simplehttpsproxy`, routing `teleport.dev.aws.stevexin.me:443` to machine1's private ip
- machine 3: tested `tsh` and `teleport start` with `HTTPS_PROXY=<domain-of-machine2>`

### CODE REVIEW

**Review by webvictim — COMMENTED** [review]
LGTM but I'm not the one you want reviewing this 😅

File: api/client/alpn_conn_upgrade_test.go [hunk 1]
```diff
@@ -273,3 +312,76 @@ func mockConnUpgradeHandler(t *testing.T, upgradeType string, write []byte) http
...
+	wg.Add(2)
+	proxyConn := func(dst, src io.ReadWriteCloser) {
+		defer dst.Close()
+		defer src.Close()
+		io.Copy(dst, src)
+		wg.Done()
+	}
+	go proxyConn(serverConn, clientConn)
+	go proxyConn(clientConn, serverConn)
+	wg.Wait()
```
[the remark this answers is not in the document]
Comment (greedy52) [reply api/client/alpn_conn_upgrade_test.go:366]: Good catch! Didn't know we have one already. The implementation is almost identical. Let me find a good place to share them.

File: lib/srv/alpnproxy/simplehttpsproxy/main.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func main() {
```
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/srv/alpnproxy/simplehttpsproxy/main.go:32]: @smallinsky I don't think `./examples` is a good fit as the program does not demo any Teleport stuff.

What do you think about moving this to `./assets/` or `./integrations/`? Alternatively, I can drop this program from this change and create a personal repo for this program importing teleport repo.

As for sharing the `ProxyHandler` from integration test, what do you think moving it to `./api/testhelpers`?

[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/srv/alpnproxy/simplehttpsproxy/main.go]: I think that ./assets/` should work.

> As for sharing the ProxyHandler from integration test, what do you think moving it to ./api/testhelpers?
👍

**Review by smallinsky — APPROVED** [review]
LGMT when remaining comments will be addressed.

File: api/client/alpn_conn_upgrade.go [hunk 3]
```diff
@@ -48,19 +53,28 @@ import (
...
 		return result
 	}
 
-	netDialer := &net.Dialer{
-		Timeout: defaults.DefaultIOTimeout,
-	}
+	// Use NewDialer which takes care of ProxyURL, and use a shorter I/O
+	// timeout to avoid blocking caller.
+	baseDialer := NewDialer(
+		context.Background(),
```
[the remark this answers is not in the document]
Comment (greedy52) [reply api/client/alpn_conn_upgrade.go:64]: Yes it should have the context passed in. To be honest i don't know why it doesn't have the context in the first place. will update.

[the remark this answers is not in the document]
Comment (greedy52) [reply api/client/alpn_conn_upgrade.go]: Added in https://github.com/gravitational/teleport/pull/27583/commits/eacc8e23c3da87ea91b60d340bf8e0a952e7d3f8

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**